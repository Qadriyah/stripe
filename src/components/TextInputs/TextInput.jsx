import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { attachToParentComponent } from './helper';
import { isEmpty } from '../../utils';

class TextInput extends Component {
  state = {
    value: '',
    clientError: ''
  };

  textInputRef = React.createRef();

  componentDidMount() {
    attachToParentComponent(this);
  }

  /**
   * Handles input value change
   */
  handleChange = event => {
    this.setState({
      value: event.target.value,
      clientError: ''
    });
  };

  /**
   * Gets the input value
   *
   * @returns string
   */
  getValue = () => {
    const { value } = this.state;
    return value;
  };

  /**
   * Manages the focus of the input element
   */
  focus = () => this.textInputRef.current.focus();

  blur = () => {
    const { value } = this.state;
    const currentNode = this.textInputRef.current || null;
    if (currentNode.required && isEmpty(value)) {
      this.setState({
        clientError:
          currentNode && currentNode.required ? 'This field is required' : ''
      });
    }
  };

  /**
   * Renders the input field label
   *
   * @param string label
   * @param string name
   * @returns Object label|null
   */
  renderLable = (label, name) =>
    !isEmpty(label) ? (
      // eslint-disable-next-line jsx-a11y/label-has-for
      <label htmlFor={name}>{label}</label>
    ) : null;

  /**
   * Gets the properties of the input field from the component props
   * @returns Object
   */
  getTextInputProps = () => {
    const {
      name,
      placeholder,
      type,
      pattern,
      className,
      serverError,
      label,
      required
    } = this.props;
    return {
      name,
      placeholder,
      type,
      pattern,
      className,
      serverError,
      label,
      required
    };
  };

  render() {
    const { value, clientError } = this.state;
    const props = this.getTextInputProps();

    const error = clientError || props.serverError;

    return (
      <div className="form-group">
        {this.renderLable(props.label, props.name)}
        <input
          id={props.name}
          name={props.name}
          placeholder={props.placeholder}
          value={value}
          type={props.type}
          onChange={this.handleChange}
          onFocus={this.focus}
          onBlur={this.blur}
          pattern={props.pattern}
          ref={this.textInputRef}
          required={props.required}
          className={classnames(props.className, {
            'is-invalid': error
          })}
        />
        {error && <div className="invalid-feedback">{error}</div>}
      </div>
    );
  }
}

TextInput.propTypes = {
  name: PropTypes.string.isRequired,
  placeholder: PropTypes.string.isRequired,
  label: PropTypes.string,
  type: PropTypes.string,
  pattern: PropTypes.string,
  className: PropTypes.string,
  serverError: PropTypes.string,
  required: PropTypes.bool
};

TextInput.defaultProps = {
  type: 'text',
  pattern: null,
  label: '',
  required: false,
  className: 'form-control form-control-lg',
  serverError: ''
};

export default TextInput;
