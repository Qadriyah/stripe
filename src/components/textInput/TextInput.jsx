import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { attachToParentComponent } from './helper';

class TextInput extends Component {
    constructor(props) {
        super(props);
        this.state = {
            value: '',
            clientError: ''
        };
        this.textInputRef = React.createRef();
    }

    componentDidMount() {
        attachToParentComponent(this);
    }

    /**
     * Handles input value change
     */
    handleChange = event => this.setState({ value: event.target.value });

    /**
     * Gets the input value
     * 
     * @returns string
     */
    getValue = () => {
        const { value } = this.state;
        return value;
    }

    /**
     * Manages the focus of the input element
     */
    focus = () => this.textInputRef.current.focus();

    render() {
        const { value, clientError } = this.state;
        const { name, placeholder, type, pattern, className, serverError, label } = this.props;
        const error = clientError || serverError;

        return (
            <div className="form-group">
                <label htmlFor={name}>{label.toUpperCase()}</label>
                <input
                    id={name}
                    name={name}
                    placeholder={placeholder}
                    value={value}
                    type={type}
                    onChange={this.handleChange}
                    onClick={this.focus}
                    pattern={pattern}
                    ref={this.textInputRef}
                    className={classnames(className, {
                        "is-invalid": error
                    })}
                />
                {error && <div className="invalid-feedback">{error}</div>}
            </div>
        )
    }
}

TextInput.propTypes = {
    name: PropTypes.string.isRequired,
    placeholder: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    type: PropTypes.string,
    pattern: PropTypes.string,
    className: PropTypes.string.isRequired,
    serverError: PropTypes.string.isRequired,
}

TextInput.defaultProps = {
    type: 'text',
    pattern: null
}

export default TextInput;
