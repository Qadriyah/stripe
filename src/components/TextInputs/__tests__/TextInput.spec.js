import React from 'react';
import { mount, shallow } from 'enzyme';
import TextInput from '../TextInput';

describe('TextInput', () => {
  const props = {
    name: 'name',
    placeholder: 'Full name',
    type: 'text',
    pattern: null,
    className: 'form-control form-control-lg',
    serverError: '',
    label: 'Name on card',
    required: true,
    parent: {
      setState: jest.fn(),
      state: {}
    }
  };

  const setup = (shouldMount = false) => {
    const wrapper = shouldMount
      ? mount(<TextInput {...props} />)
      : shallow(<TextInput {...props} />);
    return wrapper;
  };

  test('renders without failures', () => {
    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
  });

  test('it should focus/blur the input element', () => {
    const wrapper = setup(true);
    wrapper.instance().focus = jest.fn();
    wrapper.instance().blur = jest.fn();
    wrapper.update();

    wrapper.find('input').simulate('focus');
    wrapper.find('input').simulate('blur');
    expect(wrapper.instance().focus).toBeTruthy();
    expect(wrapper.instance().blur).toBeTruthy();
  });
});
