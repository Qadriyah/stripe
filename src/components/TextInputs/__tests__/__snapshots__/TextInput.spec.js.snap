// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput renders without failures 1`] = `
<div
  className="form-group"
>
  <label
    htmlFor="name"
  >
    Name on card
  </label>
  <input
    className="form-control form-control-lg"
    id="name"
    name="name"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    pattern={null}
    placeholder="Full name"
    required={true}
    type="text"
    value=""
  />
</div>
`;
